---
title: "Epigenetic Drift Age Analysis"
author: "William Casazza"
date: '2019-08-14'
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(reticulate)
library(R.matlab)
library(RSpectra)
use_condaenv("tf_vae")
knitr::opts_chunk$set(echo = TRUE)
```

# Goals
* Analyze the extent to which epigenetic drift seems to be happening in rosmap cohort 
    * global change in methylation
    * change in methylation across individual probes with subject age
* analyze the extent to which we observe more instances of mediation when we restrict ages to different groups and bins
    * Simple way is bins
    * More robust and perhaps better powered way would be to use DCARS

## Load in methylation and phenotype data
```{r}
require(tidyverse)
require(data.table)
require(RSpectra)
require(R.matlab)
methy_data <- data.table::fread("tar zxfO ~/ROSMAPmethylationWAgeSex.tar.gz", header = T)
tmp_data <- scale(t(as.matrix(methy_data[,2:ncol(methy_data)])))
decomp <- RSpectra::svds(tmp_data,10)
probe_names <- methy_data$V1
rownames(methy_data) <- probe_names
phen_data <- read.csv("~/ROSMAP_PHEN.csv", row.names = 1)
mask <- match(colnames(methy_data), rownames(phen_data))
phen_data <- phen_data[rownames(phen_data) %in% colnames(methy_data), ]
methy_data <- methy_data[,rownames(phen_data), with=FALSE]
methy_data <- resid(lm(t(as.matrix(methy_data)) ~ phen_data$msex))
exprs <- readMat("~/expressionNonAgeSexAdj.mat")
expression <- exprs$Res[[1]]
rownames(expression) <- unlist(exprs$Res[[2]])
colnames(expression) <- unlist(exprs$Res[[3]])
phen <- readMat("~/phenotype.mat")
phenotype <- phen$Pm[[1]]
rownames(phenotype) <- unlist(phen$Pm[[2]])
colnames(phenotype) <- unlist(phen$Pm[[3]])
# regress out covariates
decomp <- svds(expression,10)
expression <- expression - (decomp$u %*% diag(decomp$d) %*% t(decomp$v))
expression <- expression[match(rownames(phenotype), rownames(expression)),]
all(rownames(expression) == rownames(phenotype))
fit <- lm(as.matrix(expression) ~  phenotype[,"msex"])
expression <- resid(fit)
```


## Global Changes in methylation
```{r}
mean_methy<- rowMeans(methy_data)
fit <- lm(mean_methy ~ phen_data$age.death + phen_data$msex)
plot(fit)
summary(fit)
plot(phen_data$age.death, mean_methy)
abline(coef(fit)[c(1,2)])


fit2 <- glm(parkdx ~ mean_methy + age.death + msex, family = "binomial", data = na.omit(data.frame(msex = phen_data$msex, parkdx = phen_data$parkdx,age.death = phen_data$age.death, mean_methy = mean_methy)))
# plot(fit2)
summary(fit2)
plot(phen_data$parkdx, mean_methy)
abline(coef(fit2)[c(1,2)])
fit2 <- glm(cAD ~ mean_methy + age.death + msex,family = "binomial", data = na.omit(data.frame(msex = phen_data$msex, cAD = phen_data$cAD,age.death = phen_data$age.death, mean_methy = mean_methy)))
# plot(fit2)
summary(fit2)
plot(phen_data$parkdx, mean_methy)
abline(coef(fit2)[c(1,2)])
```



## Changes in methylation across subjects
```{r}
res <- suppressWarnings(apply(methy_data,1, function(x) cor.test(x,phen_data$age.death, method = "spearman")))
```

```{r}
res_df <- as.data.frame(t(sapply(res, function(x) c(estimate = x$estimate, pval = x$p.value))))
res_df <- res_df %>% mutate(pbonf = p.adjust(pval, method = "bonf"))
sum(res_df$pbonf < 0.05)
```
```{r}
tmp <- sapply(res, function(x) x$p.value)
df <- data.frame(p =tmp, p_adj = p.adjust(tmp, method = "bonf"))
```
## Mediation at different ages (implemented in drift_experiment.py)

Loading in data from each quantile:
```{r}
data_dir <- "~/vaecit/data/epigenetic_drift_experiment/"
fmt <- ".*_([0-3])_quantile_(.*)_5.*.csv"
result <- list()
for(f in dir(data_dir, pattern = "^_.*")){
  df <- read.csv(paste0(data_dir,f))
  df$quant <- as.numeric(gsub(fmt,"\\1",f))
  df$method <- gsub(fmt,"\\2",f)
  rev_df <- read.csv(paste0(data_dir,"rev",f))
  colnames(rev_df) <- paste0("rev_",colnames(rev_df))
  result[[f]] <- cbind(df, rev_df)
}

drift_df <- bind_rows(result)
```


```{r, fig.width=12, fig.height=5}
alpha <- 0.05
classify_p_value <- function(x,y,n){
      ifelse(x < (alpha / n) & y > (alpha / n),
        "Epigenetic Mediation",
        ifelse(x > (alpha / n) & y < (alpha/ n),
          "Transcriptional Mediation",
          ifelse(x > (alpha / n) & y > (alpha/ n),
            "Independent Association",
            "Unclassified")))
    
}

classified_drift <- drift_df %>% group_by(quant) %>% mutate(mediation = classify_p_value(omni_p, rev_omni_p, n())) %>% ungroup()
to_plot <- classified_drift %>% group_by(quant,method,mediation) %>% summarize(num_classified = n()) %>% ungroup() %>%  filter(mediation != "Independent Association")# %>% mutate(mediation = gsub("Mediation", "", mediation), method = factor(method, levels = c("fastica","pca","lfa","kernelpca",unique(method)[c(4,5)]))) 
  ggplot(to_plot, aes(x= quant, y= num_classified, fill = mediation)) + geom_bar(stat = "identity") + facet_wrap(mediation~method, nrow=2)
  ggplot(to_plot, aes(x= quant, y= num_classified, fill = mediation)) + geom_bar(stat = "identity") + facet_wrap(~method, nrow=1)
```
## Mediation at different ages (ignoring multivariate stuff for now)

```{r}
manifest <- read.csv("../CIT.txt", sep= "\t", stringsAsFactors = T)
epi_less <- read.csv("~/vaecit/scripts/_leMean_5_cit_replication_perm_test.csv")
epi_less$gene <- manifest$gene
epi_greater <- read.csv("~/vaecit/scripts/_geMean_5_cit_replication_perm_test.csv")
epi_greater$gene <- manifest$gene
(sig_less <- epi_less %>% filter(omni_p * n() < 0.05))
(sig_greater <- epi_greater %>% filter(omni_p * n() < 0.05))
to_plot <- data.frame(group = c("Younger", "Older"), amt_Mediation=c(604,805)) %>% mutate(group = factor(group, levels = c("Younger", "Older")))
ggplot(to_plot, aes(x=group, y = amt_Mediation)) + geom_bar(stat = "identity") + ggtitle("Epigenetic Mediation in Young Versus Old (single SNPs)") + theme_minimal()
```
## DCARS
Input
  * Expression and methylation rankings
    * sort along columns by age, compute rank of each methylation probe
```{r}
eQTX_manifest <- read.delim("/home/wcasazza/eQTX_manifest.txt", sep = " ")
expression <- expression[rownames(expression) %in% rownames(methy_data),]
#methy_data <- methy_data[match(rownames(methy_data),rownames(expression))]
all(rownames(expression) == rownames(methy_data))
all(rownames(expression) == rownames(phen_data))
express_ranks <- apply(t(expression[order(phen_data$age.death),]),1,rank)
methy_ranks <- apply(t(methy_data[order(phen_data$age.death),]),1,rank)
colnames(methy_ranks) <- probe_names
edges<- apply(eQTX_manifest,1,function(x) strsplit(x[3],","))
names(edges) <- eQTX_manifest$gene
edges <- as.matrix(bind_rows(edges,.id = "gene"))
all_data <- rbind(t(express_ranks),t(methy_ranks))
```


```{r}
suppressMessages(result <- DCARS::DCARSacrossNetwork(all_data,edges[c(1,2),], extractPermutationTestStatistics = T, niter=1000))
saveRDS(result, "DCARS_network.RDS")
```
```{r}
methy_data <- data.table::fread("tar zxfO /media/wcasazza/DATA2/wcasazza/ROSMAP/ROSMAPmethylationWAgeSex.tar.gz", header = T)
tmp_data <- scale(t(as.matrix(methy_data[,2:ncol(methy_data)])))
decomp <- RSpectra::svds(tmp_data,10)
probe_names <- methy_data$V1
rownames(methy_data) <- probe_names
phen_data <- read.csv("/media/wcasazza/DATA2/wcasazza/ROSMAP/ROSMAP_PHEN.csv", row.names = 1)
mask <- match(colnames(methy_data), rownames(phen_data))
phen_data <- phen_data[rownames(phen_data) %in% colnames(methy_data), ]
methy_data <- methy_data[,rownames(phen_data), with=FALSE]
methy_data <- resid(lm(t(as.matrix(methy_data)) ~ phen_data$msex))
exprs <- readMat("/media/wcasazza/DATA2/wcasazza/ROSMAP/expressionNonAgeSexAdj.mat")
expression <- exprs$Res[[1]]
rownames(expression) <- unlist(exprs$Res[[2]])
colnames(expression) <- unlist(exprs$Res[[3]])
phen <- readMat("/media/wcasazza/DATA2/wcasazza/ROSMAP/phenotype.mat")
phenotype <- phen$Pm[[1]]
rownames(phenotype) <- unlist(phen$Pm[[2]])
colnames(phenotype) <- unlist(phen$Pm[[3]])
# regress out covariates
decomp <- svds(expression,10)
expression <- expression - (decomp$u %*% diag(decomp$d) %*% t(decomp$v))
expression <- expression[match(rownames(phenotype), rownames(expression)),]
all(rownames(expression) == rownames(phenotype))
fit <- lm(as.matrix(expression) ~  phenotype[,"msex"])
expression <- resid(fit)

eQTX_manifest <- read.delim("/home/wcasazza/eQTX_manifest.txt", sep = " ")
expression <- expression[rownames(expression) %in% rownames(methy_data),]
#methy_data <- methy_data[match(rownames(methy_data),rownames(expression))]
all(rownames(expression) == rownames(methy_data))
all(rownames(expression) == rownames(phen_data))
express_ranks <- apply(t(expression[order(phen_data$age.death),]),1,rank)
methy_ranks <- apply(t(methy_data[order(phen_data$age.death),]),1,rank)
colnames(methy_ranks) <- probe_names
edges<- apply(eQTX_manifest,1,function(x) strsplit(x[3],","))
names(edges) <- eQTX_manifest$gene
edges <- as.matrix(bind_rows(edges,.id = "gene"))
all_data <- rbind(t(express_ranks),t(methy_ranks))

globalCors = apply(edges, 1, function(x) cor(all_data[x[1],], all_data[x[2],]))
stats <- readRDS("../scripts/DCARS_network_teststats.RDS")
head(stats)
sampleindices = DCARS::stratifiedSample(abs(globalCors), length = 50)
W <- DCARS::weightMatrix(ncol(all_data), type = "triangular", span = 0.5, plot = FALSE)
permstats <- DCARS::DCARSacrossNetwork(all_data,
                               edgelist = edges[sampleindices,],
                               W = W, 
                               niter = 1000,
                               weightedConcordanceFunction = DCARS::weightedPearson_matrix,
                               weightedConcordanceFunctionW = "matrix",
                               verbose = FALSE,
                               extractPermutationTestStatistics = TRUE)
pvals <- DCARS::estimatePvaluesSpearman(stats =stats, 
                               globalCors = globalCors,
                               permstats = permstats,
                               usenperm = T, nperm = 5000, verbose = F)

```
```{r}
saveRDS(pvals, "DCARS_pvals.RDS")
plot(stats, -log10(pvals$pval), col = "red", pch = 16)
pval_FDR = p.adjust(pvals$pval, method ="BH")
FDR_sig = pval_FDR < (0.5)
sum(pval_FDR < 0.5)
points(stats[FDR_sig], -log10(pvals$pval)[FDR_sig], col = "blue", pch = 16, cex = 1.2)
```

### Notes on DCARS
So, restricting to QTMs has nothign leads to pretty low detection, perhaps I should restrict to probes in regulatory regions and start from there.
```{r}
methy_data <- data.table::fread("tar zxfO /media/wcasazza/DATA2/wcasazza/ROSMAP/ROSMAPmethylationWAgeSex.tar.gz", header = T)
tmp_data <- scale(t(as.matrix(methy_data[,2:ncol(methy_data)])))
decomp <- RSpectra::svds(tmp_data,10)
probe_names <- methy_data$V1
rownames(methy_data) <- probe_names
phen_data <- read.csv("/media/wcasazza/DATA2/wcasazza/ROSMAP/ROSMAP_PHEN.csv", row.names = 1)
mask <- match(colnames(methy_data), rownames(phen_data))
phen_data <- phen_data[rownames(phen_data) %in% colnames(methy_data), ]
methy_data <- methy_data[,rownames(phen_data), with=FALSE]
methy_data <- resid(lm(t(as.matrix(methy_data)) ~ phen_data$msex))
exprs <- readMat("/media/wcasazza/DATA2/wcasazza/ROSMAP/expressionNonAgeSexAdj.mat")
expression <- exprs$Res[[1]]
rownames(expression) <- unlist(exprs$Res[[2]])
colnames(expression) <- unlist(exprs$Res[[3]])
phen <- readMat("/media/wcasazza/DATA2/wcasazza/ROSMAP/phenotype.mat")
phenotype <- phen$Pm[[1]]
rownames(phenotype) <- unlist(phen$Pm[[2]])
colnames(phenotype) <- unlist(phen$Pm[[3]])
# regress out covariates
decomp <- svds(expression,10)
expression <- expression - (decomp$u %*% diag(decomp$d) %*% t(decomp$v))
expression <- expression[match(rownames(phenotype), rownames(expression)),]
all(rownames(expression) == rownames(phenotype))
fit <- lm(as.matrix(expression) ~  phenotype[,"msex"])
expression <- resid(fit)

eQTX_manifest <- read.delim("/media/wcasazza/DATA2/wcasazza/ROSMAP/masterCpGAnnotation.tsv")#read.delim("/home/wcasazza/eQTX_manifest.txt", sep = " ")
expression <- expression[rownames(expression) %in% rownames(methy_data),]
all(rownames(expression) == rownames(methy_data))
all(rownames(expression) == rownames(phen_data))
express_ranks <- apply(t(expression[order(phen_data$age.death),]),1,rank)
methy_ranks <- apply(t(methy_data[order(phen_data$age.death),]),1,rank)
colnames(methy_ranks) <- probe_names

edges <- eQTX_manifest[,c("RefGene","TargetID")]
to_keep <- edges$RefGene %in% rownames(all_data) & edges$TargetID %in% rownames(all_data)
edges <- edges[to_keep,]
all_data <- rbind(t(express_ranks),t(methy_ranks))

globalCors = apply(edges, 1, function(x) cor(all_data[x[1],], all_data[x[2],]))
stats <- readRDS("../scripts/DCARS_network_reg_regions_teststats.RDS")[to_keep]
sampleindices = DCARS::stratifiedSample(abs(globalCors), length = 50)
W <- DCARS::weightMatrix(ncol(all_data), type = "triangular", span = 0.5, plot = FALSE)
permstats <- DCARS::DCARSacrossNetwork(all_data,
                               edgelist = edges[sampleindices,],
                               W = W, 
                               niter = 1000,
                               weightedConcordanceFunction = DCARS::weightedPearson_matrix,
                               weightedConcordanceFunctionW = "matrix",
                               verbose = FALSE,
                               extractPermutationTestStatistics = TRUE)
pvals <- DCARS::estimatePvaluesSpearman(stats =stats, 
                               globalCors = globalCors,
                               permstats = permstats,
                               usenperm = T, nperm = 5000, verbose = F)
saveRDS(pvals, "DCARS_pvals_reg_regions.RDS")
```
```{r}

plot(stats, -log10(pvals$pval), col = "red", pch = 16)
pval_FDR = p.adjust(pvals$pval, method ="BH")
FDR_sig = pval_FDR < (0.5)
sum(pval_FDR < 0.5)
points(stats[FDR_sig], -log10(pvals$pval)[FDR_sig], col = "blue", pch = 16, cex = 1.2)
```
